### PR TITLE
Fix: Virtual Services with overlapping SNI domains throws error in Gloo

### PR DESCRIPTION
**Pull request by mlholland**
# Description

Conflicting Virtualservices that have overlapping SNI domains within the same gateway cause an Envoy LDS failure and inconsistent validation state. Attempts to apply changes like this should be rejected by gloo before reaching envoy.

# Context

https://github.com/solo-io/gloo/issues/3459

# Checklist:

- [X] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [X] If I updated APIs (our protos) or helm values, I ran `make install-go-tools generated-code` to ensure there will be no code diff
- [X] I followed guidelines laid out in the Gloo [contribution guide](https://docs.solo.io/gloo/latest/contributing/)
- [X] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/3459

### CODE REVIEW

File: projects/gateway/pkg/translator/http.go [hunk 1]
```diff
@@ -28,6 +28,10 @@ var (
 		return errors.Errorf("domain conflict: the [%s] domain is present in other virtual services "+
 			"that belong to the same Gateway as this one: %v", domain, conflictingVsNames)
 	}
+	SniDomainInOtherVirtualServicesErr = func(domain string, conflictingVsNames []string) error {
```
Comment (kdorosh) [inline projects/gateway/pkg/translator/http.go:31]: vsNames probably won't be enough, we will want vsRefs (`namespace.name` for each vs)

File: projects/gateway/pkg/translator/http.go [hunk 2]
```diff
@@ -108,6 +113,54 @@ func validateVirtualServiceDomains(gateway *v1.Gateway, virtualServices v1.Virtu
...
+	}
+
+	var conflictingSniDomains []string
+	for sniDomain, vsWithThisSniDomain := range vsBySniDomain {
+		if len(vsWithThisSniDomain) > 1 {
+			conflictingSniDomains = append(conflictingSniDomains, sniDomain)
+			for i, vs := range vsWithThisSniDomain {
+				var conflictingVsNames []string
+				for j, otherVs := range vsWithThisSniDomain {
+					if i != j {
```
Comment (kdorosh) [inline projects/gateway/pkg/translator/http.go:151]: let's use better names than `i` and `j`

File: projects/gateway/pkg/translator/http.go (hunk 1 above)
Comment (kdorosh) [reply projects/gateway/pkg/translator/http.go:31]: looks like you are using the right format here, but the name is misleading and should be changed

File: projects/gateway/pkg/translator/http.go [hunk 3]
```diff
@@ -108,6 +113,54 @@ func validateVirtualServiceDomains(gateway *v1.Gateway, virtualServices v1.Virtu
...
+
+		// Add warning and skip if no virtual host
+		if vs.VirtualHost == nil {
+			reports.AddWarning(vs, NoVirtualHostErr(vs).Error())
+			continue
+		}
+
+		// todo question: is this still a concern for sni domains?
+		// Not specifying any sniDomains is not an error per se, but we need to check whether multiple virtual services
+		// don't specify any, so we use the empty string as a placeholder in this function.
```
Comment (kdorosh) [inline projects/gateway/pkg/translator/http.go:133]: we should allow multiple VS to specify SSL configs with empty SNI domains without any error

File: projects/gateway/pkg/translator/http.go (hunk 1 above)
Comment (mlholland) [reply projects/gateway/pkg/translator/http.go:31]: Yeah kept the live input the same, changed the variable name within the error function.

File: projects/gateway/pkg/translator/http.go (hunk 2 above)
Comment (mlholland) [reply projects/gateway/pkg/translator/http.go]: Changed to idx1/2.

File: projects/gateway/pkg/translator/http.go (hunk 3 above)
Comment (mlholland) [reply projects/gateway/pkg/translator/http.go]: Ah great to know, will stop changing empty domains to an empty string for checking purposes.

File: projects/gateway/pkg/translator/http.go [hunk 4]
```diff
@@ -108,6 +113,48 @@ func validateVirtualServiceDomains(gateway *v1.Gateway, virtualServices v1.Virtu
...
+	vsBySniDomain := map[string]v1.VirtualServiceList{}
+	for _, vs := range virtualServices {
+
+		// Add warning and skip if no virtual host
+		if vs.VirtualHost == nil {
+			reports.AddWarning(vs, NoVirtualHostErr(vs).Error())
+			continue
+		}
+
+		if vs.SslConfig != nil && vs.SslConfig.SniDomains != nil {
```
Comment (kdorosh) [inline projects/gateway/pkg/translator/http.go:130]: nit,
```golang
len(vs.GetSslConfig().GetSniDomains()) > 0
```
would be slightly preferred here, as it's a bit simpler imo

File: projects/gloo/pkg/translator/translator_test.go [hunk 5]
```diff
@@ -1836,6 +1836,24 @@ var _ = Describe("Translator", func() {
...
+						},
+						{
+							SniDomains: []string{"a.com"},
+						},
+					},
+				}
+				report := &validation.ListenerReport{}
+				ValidateListenerSniDomains(listener, report)
+				Expect(report.Errors).NotTo(BeNil())
+				Expect(len(report.Errors)).To(Equal(1))
```
Comment (kdorosh) [inline projects/gloo/pkg/translator/translator_test.go:1854]: just fyi, gomega has a `HaveLen()` matcher

**Pull request merged**